api/v1alpha1/httproute_types.go

```diff
@@ -190,6 +200,75 @@ type HTTPRouteAction struct {
 	Extension *core.TypedLocalObjectReference `json:"extension"`
 }
 
+// TLSConfig defines the schema of a TLS configuration.
+type TLSConfig struct {
+	// Termination defines how to terminate TLS connections.
+	//
+	// If unspecified, TLS termination type "Edge" will be used.
```

Comment:
A significant number of users don't understand the details of TLS and higher-level protocols like HTTP and hence they associate TLS with a specific route, which is incorrect, since TLS session is established first and then the route is selected based on metadata of the higher-level protocol.

And then, there is the case were two different HTTPRoute (or other types of Route) have the same SNI but different cert. Sometimes we see that there are two secrets (no cross-namespace referencing of secret) but the content it same, other times, they will have different content as well. We usually honor the oldest Ingress resource in that case usually. 

Here, we should be explicit about the behavior. I'm not sure what the solution should be but it should be hard to break production setups easily.

Reply:
> A significant number of users don't understand the details of TLS and higher-level protocols like HTTP and hence they associate TLS with a specific route, which is incorrect, since TLS session is established first and then the route is selected based on metadata of the higher-level protocol.

TLS cannot decide whether it is transport or application layer and has some unfortunate metadata redundancy with respect to HTTP.  Does the API need to reflect that?

> And then, there is the case were two different HTTPRoute (or other types of Route) have the same SNI but different cert. Sometimes we see that there are two secrets (no cross-namespace referencing of secret) but the content it same, other times, they will have different content as well. We usually honor the oldest Ingress resource in that case usually.

You could have two certificates with overlapping CNs or SANs in a `Gateway` object's `spec.listeners.tls.certificates` specification.  I am not confident the API can be designed in such a way that it cannot be easily and accidentally used to express certificates that conflict with each other; it will be necessary to detect and report conflicts at run time.

> Here, we should be explicit about the behavior. I'm not sure what the solution should be but it should be hard to break production setups easily.

Report conflicts in a status condition?

Reply:
> A significant number of users don't understand the details of TLS and higher-level protocols like HTTP and hence they associate TLS with a specific route, which is incorrect, since TLS session is established first and then the route is selected based on metadata of the higher-level protocol.

IMHO, the default common case should be to require the SNI and HTTP names to match (see #72).

Reply:
> TLS cannot decide whether it is transport or application layer and has some unfortunate metadata redundancy with respect to HTTP. Does the API need to reflect that?

No. I'm advocating to have TLS not be associated with our HTTP metadata based resources like `HTTPRoute`. That removes the possibility of expressing a certificate associated with a specific route (and that is not possible to enforce because of how TLS and HTTP work).

Reply:
@hbagdi A couple of times I started writing a response here, but I'm not sure I understand your position well enough :)

WDYT about filing a separate issue to capture a user story for the case you are advocating?